### PR TITLE
adding relative & absolute mip gaps

### PR DESCRIPTION
**Pull request by zengjian-hu-rai**
https://relationalai.atlassian.net/jira/software/c/projects/RAI/boards/41?search=jian&selectedIssue=RAI-17621

Added supports for _relative_gap_tolerance_ and _absolute_gap_tolerance_, and the report of final _relative_gap_ from solver. Note that MOI doesn't support reporting the final absolute gap.

### CODE REVIEW

File: test/all_tests.jl [hunk 1]
```diff
@@ -81,6 +81,8 @@ end
     # solve and check output is expected for each input json file
     @testset "$j" for j in json_names
         output = JSON3.read(run_solve(read_json("inputs", j)))
+        @info output
+        #isdefined(Main, :Infiltrator) && Main.infiltrate(@__MODULE__, Base.@locals, @__FILE__, @__LINE__)
```
Comment (chriscoey) [inline test/all_tests.jl:85]: remove comment

Comment (chriscoey) [reply test/all_tests.jl:85]: and the `@info`

Comment (zengjian-hu-rai) [reply test/all_tests.jl:85]: oh my goodness. thought i 've removed them. :)

File: src/SolverAPI.jl [hunk 2]
```diff
@@ -111,6 +111,16 @@ function response(
     res["solve_time_sec"] = Float64(MOI.get(model, MOI.SolveTimeSec()))
 
     result_count = MOI.get(model, MOI.ResultCount())
+
+    try
+        res["relative_gap"] = Float64(MOI.get(model, MOI.RelativeGap()))
```
Comment (chriscoey) [inline src/SolverAPI.jl:116]: makes a little more sense to do 
```
rel_gap = Float64(MOI.get(model, MOI.RelativeGap()))
``` 
and then only set `res["relative_gap"]` if it's finite.

File: test/all_tests.jl (hunk 1 above)
Comment (zengjian-hu-rai) [reply test/all_tests.jl]: done

**Review by chriscoey — APPROVED** [review]

File: src/SolverAPI.jl (hunk 2 above)
Comment (zengjian-hu-rai) [reply src/SolverAPI.jl]: thanks. good point

File: src/SolverAPI.jl [hunk 3]
```diff
@@ -113,9 +113,10 @@ function response(
     result_count = MOI.get(model, MOI.ResultCount())
 
     try
-        res["relative_gap"] = Float64(MOI.get(model, MOI.RelativeGap()))
-        if isinf(res["relative_gap"])
-            res["relative_gap"] = nothing # Inf cannot be serialized to JSON
+        res["relative_gap"]
```
Comment (chriscoey) [inline src/SolverAPI.jl:116]: remove this line?

File: src/SolverAPI.jl [hunk 4]
```diff
@@ -113,9 +113,10 @@ function response(
...
+        rel_gap = Float64(MOI.get(model, MOI.RelativeGap()))
+        if isinf(rel_gap)
+            res["relative_gap"] = rel_gap# Inf cannot be serialized to JSON
```
Comment (chriscoey) [inline src/SolverAPI.jl:119]:
```suggestion
            res["relative_gap"] = rel_gap # Inf cannot be serialized to JSON
```

Comment (chriscoey) [reply src/SolverAPI.jl]: yes

**Pull request merged**